File: Astromaximum/Classes/utils/AmaxDataInputStream.m
//
//  AmaxDataInputStream.m
//  Astromaximum
//
//  Created by admin on 14.11.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

#import "Astromaximum-Bridging-Header.h"

@implementation AmaxDataInputStream

- (AmaxDataInputStream *)initWithBytes:(const void *)bytes length:(NSUInteger)length
{
    dataLength = length;
    data = malloc(dataLength);
    memcpy(data, bytes, dataLength);
    return self;
}

- (AmaxDataInputStream *)initWithData:(const NSData *)rawData
{
    dataLength = [rawData length];
    data = malloc(dataLength);
    memcpy(data, [rawData bytes], dataLength);
    return self;
}

- (void) dealloc
{
    free(data);
}

- (void)reset
{
    position = 0;
}

- (void)skipBytes:(size_t)byteCount
{
    position += byteCount;
}

- (BOOL)reachedEOF
{
    return position >= dataLength;
}

- (SInt16)readShort
{
    SInt16 res = data[position++];
    res <<= 8;
    res += data[position++];
    return res;
}

- (unsigned char)readUnsignedByte
{
    return data[position++];
}

- (char)readByte
{
    return data[position++];
}

- (UInt16)readUnsignedShort
{
    UInt32 n0 = data[position++];
    UInt32 n1 = data[position++];
    return (n0 << 8) + n1;
    
}

- (int)readInt
{
    int res = data[position++];
    res <<= 8;
    res += data[position++];
    res <<= 8;
    res += data[position++];
    res <<= 8;
    res += data[position++];
    return res;
}

- (Size)availableBytes
{
    return dataLength - position;
}

- (void)readToBuffer:(void *)buffer length:(Size)byteCount
{
    memcpy(buffer, &data[position], byteCount);
    position += byteCount;
}

- (NSString *)readUTF
{
    UInt16 stringLength = [self readUnsignedShort];
    char *buffer = malloc(stringLength + 1);
    [self readToBuffer:buffer length:stringLength];
    buffer[stringLength] = 0;
    return [[NSString alloc]initWithCString:buffer encoding:NSUTF8StringEncoding];
}

@end

File: Astromaximum/Classes/utils/AmaxDataInputStream.swift
//
//  AmaxDataInputStream.swift
//  Astromaximum2
//
//  Created by Andrei Ivushkin on 12.07.2020.
//  Copyright © 2020 S&W Axis. All rights reserved.
//

import Foundation

class AmaxDataInputStream {

    var position: Int = 0
    var dataLength: Int = 0
    var data: Data

    init(bytes: UnsafeRawPointer, length: Int) {
        dataLength = length
        data = Data.init(bytes: bytes, count: length)
    }

    init(data rawData: Data) {
        dataLength = rawData.count
        data = rawData
    }

    func reset() {
        position = 0
    }

    func skipBytes(_ byteCount: Int) {
        position += byteCount
    }

    func reachedEOF() -> Bool {
        return position >= dataLength
    }

    func readShort() -> Int16 {
        var res = Int16(data[position])
        position += 1
        res <<= 8
        res += Int16(data[position])
        position += 1
        return res
    }

    func readUnsignedByte() -> UInt8 {
        let result = UInt8(data[position])
        position += 1
        return result
    }

    func readByte() -> Int8 {
        let result = Int8(bitPattern: data[position])
        position += 1
        return result
    }

    func readUnsignedShort() -> UInt16 {
        let n0 = UInt16(data[position])
        position += 1
        let n1 = UInt16(data[position])
        position += 1
        return UInt16((n0 << 8) + n1)
    }

    func readInt() -> Int {
        var res = Int(data[position])
        position += 1
        res <<= 8
        res += Int(data[position])
        position += 1
        res <<= 8
        res += Int(data[position])
        position += 1
        res <<= 8
        res += Int(data[position])
        position += 1
        return Int(res)
    }

    func availableBytes() -> Int {
        return dataLength - position
    }

    func read(length byteCount: Int) -> Data {
        let range = position ..< (position + byteCount)
        let buffer = data.subdata(in: range)
        position += byteCount
        return buffer
    }

    func readUTF() -> String? {
        let stringLength = Int(readUnsignedShort())
        let range = position ..< (position + stringLength)
        let buffer = data.subdata(in: range)
        let s = String.init(data: buffer, encoding: .utf8)
        position += stringLength
        return s
    }
}
